storage/objects/objects_test.go

```diff
@@ -631,3 +631,41 @@ func TestObjectBucketLock(t *testing.T) {
 		t.Errorf("temporary hold is not disabled")
 	}
 }
+
+func TestObjectRetention(t *testing.T) {
+	tc := testutil.SystemTest(t)
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		t.Fatalf("storage.NewClient: %v", err)
+	}
+	t.Cleanup(func() { client.Close() })
+
+	var (
+		bucketName = testutil.UniqueBucketName(testPrefix)
+		objectName = "foo.txt"
+	)
+
+	bucket := client.Bucket(bucketName).SetObjectRetention(true)
+	if err := bucket.Create(ctx, tc.ProjectID, nil); err != nil {
+		t.Fatalf("Bucket(%q).Create:: %v", bucketName, err)
+	}
+	defer testutil.DeleteBucketIfExists(ctx, client, bucketName)
+
+	if err := uploadFile(io.Discard, bucketName, objectName); err != nil {
+		t.Fatalf("uploadFile(%q): %v", objectName, err)
+	}
+
+	err = setObjectRetentionPolicy(io.Discard, bucketName, objectName)
+	if err != nil {
+		t.Errorf("setObjectRetention: %v", err)
+	}
+	attrs, err := bucket.Object(objectName).Attrs(ctx)
+	if err != nil {
+		t.Errorf("object.Attrs: %v", err)
+	}
+
+	if attrs.Retention == nil {
+		t.Errorf("mismatching retention config, wanted %+v, got nil", attrs.Retention)
```

Comment:
Nit: got before want. Also we should probably check that it's the expected number of days as well (well, under that number given the test execution time).